---
title: "Enzymatic Error Correction Figures"
author: "Nathan Lubock"
date: "January, 2017"
output: pdf_document
css: scroll.css
---

# R Stuff
## Knitr Options
```{r}
knitr::opts_chunk$set(fig.width = 17.5, fig.height = 10.94, dpi=300)
knitr::opts_chunk$set(fig.path = "./paper/")
knitr::opts_chunk$set(dev='pdf')
knitr::opts_chunk$set(warning=FALSE)

# see http://stackoverflow.com/q/36230790 to scroll output
# needs to be really big to prevent wrapping from happening before the scroll bar comes up
options(width = 240)
```


## Initialization
```{r, message=F}
# plotting utils
library(scales)
library(grid)
library(gridExtra)

# data.table backend
library(dtplyr)
library(data.table)

# tidyverse!
library(stringr)
library(broom)
library(magrittr)
library(tidyverse)
```


## Style Choices
```{r, pallete}
theme_pub <- function(base_size = 13, base_family = "") {
  require(grid)
  # based on https://github.com/noamross/noamtools/blob/master/R/theme_nr.R
  # start with theme_bw and modify from there!
  theme_bw(base_size = base_size, base_family = base_family) +# %+replace%
    theme(
      # grid lines
      panel.grid.major.x = element_line(colour="#ECECEC", size=0.5, linetype=1),
      panel.grid.minor.x = element_blank(),
      panel.grid.minor.y = element_blank(),
      panel.grid.major.y = element_line(colour="#ECECEC", size=0.5, linetype=1),
      panel.background   = element_blank(),

      # axis options
      axis.ticks.y   = element_blank(),
      axis.title.x   = element_text(size=rel(2.25), vjust=0.25),
      axis.title.y   = element_text(size=rel(2.25), vjust=0.35),
      axis.text      = element_text(color="black", size=rel(1.5)),

      # legend options
      legend.title    = element_blank(),
      legend.key      = element_rect(fill="white"),
      legend.key.size = unit(1, "cm"),
      legend.text     = element_text(size=rel(2)),

      # facet options
      strip.text = element_text(size=rel(2)),

      # title options
      plot.title = element_text(size=rel(3), vjust=0.25, hjust=0.5)
      )
  }

# set the theme and brewer color
theme_set(theme_pub())
cbbPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
```

## Helper Functions
```{r, functions}
DistribUncert2 <- function(df) {
  # Takes a df with uncertain types and manually distributes them, returning a
  # df of counts ready for plotting. Since we cannot place the insertion or
  # deletion precisely, we will assign fractional counts to each position based
  # on the nature of the difference. For example, a deletion in a 'AAA' repeat
  # could be at any of the A's, so we will assign a 1/3 count to each position
  #
  # Args:
  #   df - a data frame that must have the following:
  #     ColNames: Pos, Diff, Type
  #     Type: Must contain 'UI' and 'UD'
  # Returns:
  #   df - data frame with colnames Pos, Type, Count
  require(dplyr, magrittr, stringr)
  uncert <- df %>%
    filter(Type %in% c('UI', 'UD')) %>%
    mutate(
      To=str_sub(Diff, 2, 2),
      From=str_sub(Diff, 1, 1),
      Diff=str_sub(Diff, -1),
      Type=str_sub(Type, -1),
      FracCount=1/as.numeric(From)
    ) %>%
    select(c(-From, -To))

  # filter the canonical types and summarize their counts as well
  canon <- df %>%
    filter(!(Type %in% c('UI', 'UD'))) %>%
    mutate(FracCount=1)

  return(bind_rows(uncert, canon))
}

LabelMaker <- function(graph, label){
  # Takes a ggplot and adds a label to the top-left corner.
  # Must be used in conjunction with grid.arrange to plot
  # See: https://stackoverflow.com/a/29863172
  # Args:
  #   graph - a ggplot
  #   label - text string to add as label
  # Returns:
  #   gtable with plot and label
  require(ggplot2, grid, gridExtra)
  myplot <- arrangeGrob(
    graph,
    top = textGrob(
      label,
      x = unit(0, 'npc'),
      y = unit(1, 'npc'),
      just = c('left', 'top'),
      gp = gpar(fontsize=32)
    )
  )

  return(myplot)
}
```

## Data Loading
We can get about at 10x speed-up by using pure `data.table`'s, however, some of `dplyr`'s functionality does not seem to behave quite right (especially joins). This is a known issue, and is being worked on.
```{r, constants, message=FALSE, warning=FALSE, results='hide'}
ref.seq <- "GCTGCCGATTTCCATAAGATGCCTCCACGTCTCCGAAGAACTACATGGTGAATGTGTGAAGGCATTTTGAACCAATCCTCGAGCAGTGTTGTATATATCG"
refCounts <- data.table(
  Char = c('A', 'T', 'G', 'C', 'N'),
  Count = c(str_count(ref.seq, 'A'),
            str_count(ref.seq, 'T'),
            str_count(ref.seq, 'G'),
            str_count(ref.seq, 'C'),
            str_count(ref.seq, 'N'))
  )

# requisite information for all treatments
# setwd("~/Projects/errorCorrect-dev/analysis/")
charCounts <- fread('zcat ./output/char-counts.txt.gz', header=T)
data <- fread('zcat ./output/errs-all-samples.csv.gz', header=T)
allSamps <- data %>%
  filter(str_sub(Sample, 1, 1) %in% c('1', '2'))

# setwd("~/Dropbox/UCLA/Kosuri/ErrorCorrect/github/errorCorrect/analysis/")
# charCounts <- fread('./data/output/char-counts.txt', header=T)
# data <- fread('./data/output/errs-all-samples.csv', header=T)
# allSamps <- data %>%
#   filter(str_sub(Sample, 1, 1) %in% c('1', '2'))

# constants for all samples
readCounts <- fread('./output/read-counts.txt', header=T)

# subset variables for easy running
nonDoped <- allSamps %>% filter(Sample == '1_nonDoped')
doped <- allSamps %>% filter(Sample == '1_DopedTemp')
```

### Review Setup
```{r, review-setup}
c2 <- 'GGGTCACGCGTAGGACATTACTCTACGGTAAGGCGACATATGGCCAGATTAATTCCCAGTGTCGTCTAAGCGGTCTATCATAAATCGTGGATGGAGGAGGAGTGTCTCGGGCCATCCTGGTTCTTGTCGACTCCATGTCATGAAATGAAGCCCACTACTGTCTTGTTGTGATGGGAGTCACCATGTTCCAGCTGTGTCTTTAACCGTGTGGCTGCGGAAC'
c3 <- 'GGTCGAGCCGGAACTGATGACGATGGCTAAATACATGTATAATGGTACATCCCTAGCAGGCATGCCCGCTTGGTCACGTGAGTAGCCCACGATATCGTTTAGAGGTAGTATAAGGGTATGCTCAGCTGCTCACTGGTATATCAGTCAGACCTACATTACCGAACTTGATGTGGCAGTCTGACCCAACTGTCCTTGCGTTGACGATTCTGGGTGCGCATCC'

review.samples <- c(
  'C2-Q5-1', 'C2-Q5-2', 'C3-Q5-1', 'C3-Q5-2',
  'C2-Taq-1', 'C2-Taq-2', 'C3-Taq-1', 'C3-Taq-2')

review.counts <- refCounts <- data.table(
  Char = rep(c('A', 'T', 'G', 'C'), 2),
  Construct = c(rep('C2', 4), rep('C3', 4)),
  Count = c(str_count(c2, 'A'), str_count(c2, 'T'), str_count(c2, 'G'), str_count(c2, 'C'),
            str_count(c3, 'A'), str_count(c3, 'T'), str_count(c3, 'G'), str_count(c3, 'C'))
)

review <- data %>%
  filter(Sample %in% review.samples)

# free up ram
rm(data)
```


# Main Figures
## Figure 2 - Error Analysis for a Standard Oligo Assembly
```{r, Figure_2, fig.width=17.5, fig.height=17.5}
#-------------------------------------------------------------------------------
# Panel 1
# Plot the position of all types of errors

# We need to make sure that any 0's are actually caught for plotting
# nonDoped %>%
#   DistribAndNorm(., 1) %>%
#   complete(Type, Pos, fill=list('Norm'=0))

positions <- nonDoped %>%
  DistribUncert2() %>%
  count(Pos, Type, wt=FracCount) %>%
  filter(Type != 'S') %>%
  ungroup() %>%
  mutate(
    Norm = n / subset(readCounts, Sample == '1_nonDoped')$Reads,
    Type = Type %>%
      factor(levels = c('M', 'I', 'D', 'P')) %>%
      recode(D = 'Single Base Deletions', I = 'Single Base Insertions',
             P = 'Multiple Base Deletions', M = 'Mismatches')
  ) %>%
  ggplot(aes(x=Pos, y=Norm)) +
  geom_point(size=3) +
  facet_wrap(~ Type, ncol=2) +
  stat_smooth(se=F) +
  labs(x = 'Position',
       y = 'Error Rate',
       title = 'Error Rate vs. Position for Error Sub-types') +
  scale_y_log10() +
  annotation_logticks(sides='l')

#-------------------------------------------------------------------------------
# Panel 1b
# Percentage of Error Subtypes

sub_type <- nonDoped %>%
    DistribUncert2() %>%
  count(Type, wt=FracCount) %>%
  mutate(
    Norm = n / sum(n) * 100,
    Type = Type %>%
      factor(levels = c('M', 'D', 'P', 'I', 'S')) %>%
      recode(M = 'MM', D = 'Del.', P = 'M. Del.', I = 'Ins.', S = 'M. Ins.')
    ) %T>%
  {arrange(., -Norm) %>% print()} %>%
  ggplot(aes(x=Type, y=Norm)) +
  geom_bar(stat='identity') +
  theme(plot.title = element_text(size=rel(2))) +
  labs(
    y = 'Percent',
    x = 'Error Type',
    title = 'Percentage of Error Sub-types'
  )

#-------------------------------------------------------------------------------
# Panel 1c
# plot the distribution of total mismatches per position
mm_freq <- nonDoped %>%
  filter(Type == 'M') %>%
  mutate(
    To = str_sub(Diff, 2, 2),
    From = str_sub(Diff, 1, 1)
  ) %>%
  count(Pos, From) %>%
  ungroup() %>%
  # left_join(rename(refCounts, From=Char), by='From') %>%
  mutate(Norm = n / subset(readCounts, Sample == '1_nonDoped')$Reads) %T>%
  { # pairwise wilcox test and print median values for paper
    group_by(., From) %>%
      summarise(med=median(Norm)) %>%
      arrange(-med) %>%
      print; # <- ; critical for . to be interpreted correctly
    with(., pairwise.wilcox.test(n, From)) %>% print
  } %>%
  ggplot(aes(x = From, y = Norm)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(position=position_jitter(w = 0.5), size=0.75, alpha=0.8) +
  # stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom = 'crossbar', width = 0.5) +
  labs(
    y = 'Error Rate',
    title = 'Mismatches per Position'
  ) +
  theme(
    axis.text.x = element_text(size=28),
    axis.title.x = element_blank(),
    plot.title = element_text(size=rel(2.25))
  ) +
  scale_y_continuous(labels = scientific_format())

#-------------------------------------------------------------------------------
# Panel 1d
# what bases are most likely mutated to
# We will normallize by the total count in each "from" group
mm_type <- nonDoped %>%
  filter(Type == 'M') %>%
  count(Pos, Diff) %>%
  ungroup() %>%
  mutate(
    Char=str_sub(Diff, 1, 1),
    Class=Diff %>%
      recode(AT='Transversion', AG='Transition', AC='Transversion',
             TA='Transversion', TG='Transversion', TC='Transition',
             GA='Transition', GT='Transversion', GC='Transversion',
             CA='Transversion', CT='Transition', CG='Transversion')
  ) %>%
  # left_join(refCounts, by='Char') %>%
  mutate(Norm = n / subset(readCounts, Sample == '1_nonDoped')$Reads) %T>%#(Count * subset(readCounts, Sample == '1_nonDoped')$Reads)) %T>%
  {# significance testing and printing for paper
    group_by(., Diff) %>%
      summarise(med=median(Norm)) %>%
      arrange(-med) %>%
      print; # <- ; critical for . to be interpreted correctly
      with(., pairwise.wilcox.test(Norm, Diff)) %>% print
  } %>%
  ggplot(aes(x = Diff, y = Norm, color=Class)) +
  geom_boxplot(outlier.shape = NA, show.legend = FALSE) +
  geom_jitter(position=position_jitter(w = 0.5), size=0.75, alpha=0.8) +
  # stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom = 'crossbar', width = 0.5, color='black') +
  labs(
    y = 'Error Rate',
    title = 'Mismatch Sub-types per Pos.'
    ) +
  theme(
    legend.position='bottom',
    legend.key.size=unit(0.75, "cm"),
    axis.title.x=element_blank(),
    axis.text.x = element_text(angle = 315, vjust=0.5),
    plot.title = element_text(size=rel(1.75))
  ) +
  scale_y_continuous(labels = scientific_format()) +
  scale_color_manual(values = c('#7b3294', '#008837')) +
  guides(colour = guide_legend(override.aes = list(size=5)))

#-------------------------------------------------------------------------------
# plot everything!
grid.arrange(
  LabelMaker(positions, 'A)'),
  arrangeGrob(
    LabelMaker(sub_type, 'B)'),
    LabelMaker(mm_freq, 'C)'),
    LabelMaker(mm_type, 'D)'),
    nrow=1),
  ncol=1,
  heights = c(1, 0.67)
  )

```

```{r, Figure_2_vals}
# pairwise test of medians for each group
nonDoped %>%
  DistribUncert2() %>%
  count(Type, Pos, wt=FracCount) %>%
  mutate(Norm = n / subset(readCounts, Sample == '1_nonDoped')$Reads) %T>%
  with(., pairwise.wilcox.test(Norm, Type) %>% print) %>%
  group_by(Type) %>%
  summarise(med=median(Norm), mean=mean(Norm))

# insertions at position 1
nonDoped %>%
  DistribUncert2() %>%
  filter(Type == 'I', Pos == 1) %>%
  count(Diff)

# differences in medians in annealing regions and outside
nonDoped %>%
    DistribUncert2() %>%
    count(Type, Pos, wt=FracCount) %>%
    mutate(Norm = n / subset(readCounts, Sample == '1_nonDoped')$Reads) %>%
    filter(Type == 'P') %>%
  mutate(Region = if_else(Pos >= 36 & Pos <=64, 'Anneal', 'No')) %T>%
  {wilcox.test(Norm ~ Region, data=.) %>% print} %>%
  group_by(Region) %>%
  summarise(med=median(Norm), IQR=IQR(Norm))
```


## Figure 3 - Error Correction and Percent Perfects
```{r, errRates}
# reduce the all of the reads down to errors/read (this may take a while...)
# join with the length of every read and fill in perfects with a 0 count
# errRate calculation from Furhman paper
errRates <- allSamps %>%
  DistribUncert2() %>%
  count(Sample, Name, wt=FracCount) %>%
  left_join(charCounts %>% filter(!Sample %in% review.samples), ., by = c('Sample', 'Name')) %>%
  replace_na(list(n=0)) %>%
  group_by(Sample) %>%
  summarise(
    errRate = mean(n * (1000/Len)),
    sem = sd(n*(1000/Len)) / sqrt(n())
  ) %>%
  left_join(readCounts, by='Sample') %>%
  mutate(
    PercentPerf = (Reads - Errs) / Reads * 100,
    Treatment = str_sub(Sample, 1, 1),
    Sample = str_sub(Sample, 3),
    # Pretty printing for figures
    Sample = Sample %>%
      recode(nonDoped='Standard Oligo', DopedTemp='Doped Oligo',
             MutS_1900nM='MutS (1900nM)', MutS_950nM='MutS (950nM)',
             T7EndoIFurhmann='T7 EndoI (Fuhrmann)', T4EndoVII='T4 EndoVII',
             T7EndoI='T7 EndoI (0U T7 Lig.)',
             `T7EndoI-e3T7Ligase`='T7 EndoI (1e3U T7 Lig.)',
             `T7EndoI-e4T7Ligase`='T7 EndoI (1e4U T7 Lig.)',
             `ErrASE-nonDoped` = 'Standard Oligo (ErrASE)')
  )
```

```{r, Figure_3}
# manual ordering for nice plots
plt.order <- c('Standard Oligo (ErrASE)', 'Standard Oligo', 'MutS (1900nM)',
           'MutS (950nM)', 'ErrASE', 'T7 EndoI (Fuhrmann)',
           'T4 EndoVII', 'Surveyor', 'T7 EndoI (0U T7 Lig.)',
           'T7 EndoI (1e3U T7 Lig.)', 'T7 EndoI (1e4U T7 Lig.)',
           'EndoV', 'Doped Oligo')

errRates %>%
  # add in 0's for doped and nonDoped
  select(c(-Errs, -Reads, -sem)) %>%
  bind_rows(
    data.table(Sample=c('Doped Oligo', 'Standard Oligo'),
               errRate=c(0.0, 0.0),
               PercentPerf=c(0.0, 0.0),
               Treatment=c('2', '2'))
  ) %>%
  gather(Metric, Value, PercentPerf, errRate) %>%
  mutate(
    Metric=if_else(Metric == 'PercentPerf',
                   "Percent Perfect Reads",
                   "Error Frequency (per kb)") %>%
      factor(levels=c("Percent Perfect Reads",
                      "Error Frequency (per kb)")),
    Sample = factor(Sample, levels = plt.order)
  ) %>%
  ggplot(aes(x=Sample, y=Value, fill=Treatment)) +
  geom_bar(stat='identity', position='dodge') +
  facet_wrap(~ Metric, nrow=2) +
  theme(
    axis.text.x=element_text(angle=315, hjust=0.15, vjust=0.90, size=rel(1.0)),
    axis.title.x=element_blank(),
    axis.title.y=element_blank(),
    legend.title=element_text(size=rel(2.25)),
    legend.position="bottom"
    ) +
  scale_y_continuous(breaks=seq(0,60,10)) +
  scale_fill_manual(name="Treatment Round:",
                      values=c("#ca0020", "#0571b0"))
```


## Figure 4 - Enzyme Preferences
```{r, enzPref}
enzPref <- allSamps %>%
  DistribUncert2() %>%
  count(Sample, Type, Pos, wt=FracCount) %>%
  ungroup() %>%
  left_join(readCounts, by = 'Sample') %>%
  mutate(
    Treatment=str_sub(Sample, 1, 1),
    Sample=str_sub(Sample, 3),
    Norm=n / Reads
  ) %>%
  # Grab the normalized data from DopedTemp for easy divide
  left_join(.,
            filter(., Sample == 'DopedTemp') %>%
              transmute(Type=Type, Pos=Pos, Dope_Norm=Norm),
            by = c('Type', 'Pos')
  ) %>%
  mutate(
    Rel_Norm = Norm / Dope_Norm,
    Fold_2 = log2(Rel_Norm),
    Fold = 2^-Fold_2,
    # pretty print for figures
    Sample = Sample %>%
      recode(nonDoped='Standard Oligo', DopedTemp='Doped Oligo',
             MutS_1900nM='MutS (1900nM)', MutS_950nM='MutS (950nM)',
             T7EndoIFurhmann='T7 EndoI (Fuhrmann)', T4EndoVII='T4 EndoVII',
             T7EndoI='T7 EndoI (0U T7 Lig.)',
             `T7EndoI-e3T7Ligase`='T7 EndoI (1e3U T7 Lig.)',
             `T7EndoI-e4T7Ligase`='T7 EndoI (1e4U T7 Lig.)',
             `ErrASE-nonDoped` = 'Standard Oligo (ErrASE)')
  )

#-------------------------------------------------------------------------------
# specific call-outs for indels and mismatches
enzPref.idm <- allSamps %>%
  DistribUncert2() %>%
  filter(Type %in% c('I', 'M', 'D')) %>%
  count(Sample, Type, Pos, Diff, wt=FracCount) %>%
  ungroup() %>%
  left_join(readCounts, by = 'Sample') %>%
  mutate(
    Treatment=str_sub(Sample, 1, 1),
    Sample=str_sub(Sample, 3),
    Norm=n / Reads
    ) %>%
  # Grab the normalized data from DopedTemp for easy divide
  left_join(.,
            filter(., Sample == 'DopedTemp') %>%
              transmute(Diff=Diff, Type=Type, Pos=Pos, Dope_Norm=Norm),
            by = c('Type', 'Pos', 'Diff')
            ) %>%
  mutate(
    Rel_Norm = Norm / Dope_Norm,
    Fold_2 = log2(Rel_Norm),
    Fold = 2^-Fold_2,
    Class= Diff %>%
      recode(AT='Transversion', AG='Transition', AC='Transversion',
             TA='Transversion', TG='Transversion', TC='Transition',
             GA='Transition', GT='Transversion', GC='Transversion',
             CA='Transversion', CT='Transition', CG='Transversion'),
    Sample = Sample %>%
      recode(nonDoped='Standard Oligo', DopedTemp='Doped Oligo',
             MutS_1900nM='MutS (1900nM)', MutS_950nM='MutS (950nM)',
             T7EndoIFurhmann='T7 EndoI (Fuhrmann)', T4EndoVII='T4 EndoVII',
             T7EndoI='T7 EndoI (0U T7 Lig.)',
             `T7EndoI-e3T7Ligase`='T7 EndoI (1e3U T7 Lig.)',
             `T7EndoI-e4T7Ligase`='T7 EndoI (1e4U T7 Lig.)',
             `ErrASE-nonDoped` = 'Standard Oligo (ErrASE)')
  )
```


```{r, Figure_4}
# order by error frequency
plt.order <- c('Standard Oligo (ErrASE)', 'Standard Oligo',
               'ErrASE', 'T7 EndoI (Fuhrmann)', 'MutS (1900nM)',
               'MutS (950nM)', 'Surveyor', 'T4 EndoVII',
               'T7 EndoI (0U T7 Lig.)', 'T7 EndoI (1e3U T7 Lig.)',
               'T7 EndoI (1e4U T7 Lig.)', 'EndoV', 'Doped Oligo')

#-------------------------------------------------------------------------------
# Panel 1
# Plot the positional distribution across enzymes for indels and mm's
pan1 <- enzPref %>%
  filter(!Sample %in% c('Doped Oligo', 'Standard Oligo', 'Standard Oligo (ErrASE)')) %>%
  mutate(
    Type = case_when(Type == 'D' | Type == 'P' ~ 'Deletions',
                     Type == 'I' | Type == 'S' ~ 'Insertions',
                     TRUE ~ 'Mismatches'),
    # factor madness for proper ordering
    Type = factor(Type, levels = c('Mismatches', 'Deletions', 'Insertions')),
    Sample = factor(Sample, levels = plt.order)
  ) %>%
  ggplot(aes(x=Sample, y=Fold_2, color=Treatment)) +
  geom_boxplot() +
  facet_wrap(~ Type, ncol = 1) +
  theme(
    legend.position='bottom',
    legend.text=element_text(size=rel(1.5)),
    legend.title=element_text(size=rel(2.25)),
    axis.text.x=element_text(angle=305, hjust=0.15, vjust=0.90, size=rel(0.85)),
    axis.title.x=element_blank()
  ) +
  guides(colour = guide_legend(override.aes = list(size=2))) +
  scale_color_manual(
    name = 'Treatment',
    values=c("#ca0020", "#0571b0")
  ) +
  labs(y='Log2-fold Error Rate Change')

#-------------------------------------------------------------------------------
# Panel 2
# Specific call outs for ErrASE/T7 Endo
pan2 <- enzPref.idm %>%
  filter(
    Type == 'M',
    Treatment == '2',
    Sample %in% c('ErrASE', 'T7 EndoI (Fuhrmann)')
  ) %>%
  mutate(Sym = Diff %>% recode(AC = 'A/T -> C/G', TG='A/T -> C/G',
                               AG = 'A/T -> G/C', TC='A/T -> G/C',
                               AT = 'A/T -> T/A', TA='A/T -> T/A',
                               CA = 'C/G -> A/T', GT='C/G -> A/T',
                               CG = 'C/G -> G/C', GC='C/G -> G/C',
                               CT = 'C/G -> T/A', GA='C/G -> T/A')
  ) %>%
  ggplot(aes(x=Sym, y=Fold_2, color=Class)) +
  geom_boxplot(
    outlier.shape = NA,
    show.legend = FALSE
  ) +
  geom_point(
    position=position_jitter(),
    size = 0.25,
    alpha = 0.8
  ) +
  facet_wrap(~ Sample, ncol=1, scales='free_y') +
  theme(
    legend.position = 'bottom',
    axis.title.x=element_blank(),
    axis.text.x = element_text(angle = 315, vjust=0.5)
  ) +
  guides(colour = guide_legend(override.aes = list(size=5))) +
  scale_color_manual(values = c('#7b3294', '#008837')) +
  labs(y='Log2-fold Error Rate Change')

grid.arrange(
  LabelMaker(pan1, 'A)'),
  LabelMaker(pan2, 'B)'),
  ncol = 2,
  widths = c(0.67, 0.33)
)
```

```{r, Figure_4_vals}
# table version of plot
enzPref %>%
  filter(!Sample %in% c('Doped Oligo', 'Standard Oligo', 'Standard Oligo ErrASE')) %>%
  mutate(
    Type = case_when(Type == 'D' | Type == 'P' ~ 'Deletions',
                     Type == 'I' | Type == 'S' ~ 'Insertions',
                     TRUE ~ 'Mismatches')
  ) %>%
  group_by(Sample, Treatment, Type) %>%
  summarise(
    Mean=mean(Fold),
    Median=median(Fold)
  ) %>%
  ungroup() %>%
  arrange(Sample, Treatment, Type)
```

### Figure 4 - Supplement
#### Sup - Mismatch Preferences
```{r, Enzyme_Prefs_MM, warning=FALSE}
enzPref.idm %>%
  filter(
    Type == 'M',
    !Sample %in% c('Doped Oligo', 'Standard Oligo (ErrASE)')
  ) %>%
  ggplot(
    aes(x=Diff,
        y=Fold_2,
        color=interaction(Treatment, Class))
  ) +
  geom_boxplot(
    outlier.shape = NA,
    show.legend = FALSE
  ) +
  geom_point(
    position=position_jitterdodge(),
    size = 0.25,
    alpha = 0.8
  ) +
  facet_wrap(~ Sample, ncol=3, scales='free_y') +
  theme(
    legend.position = 'bottom',
    axis.title.x=element_blank(),
    axis.text.x = element_text(angle = 315, vjust=0.5)
  ) +
  guides(colour = guide_legend(override.aes = list(size=5))) +
  scale_color_manual(values = c('#c2a5cf','#7b3294', '#a6dba0','#008837')) +
  labs(
    title='Mismatch Reduction Relative to Doped Assembly',
    y='Log2-fold Error Rate Change'
  )
```

There appears to be some real differences here, let's do some testing.

```{r, mm_pref_tests}
# first run anova to make sure there are diffs in medians
enzPref.idm %>%
  filter(
    Type %in% c('D', 'I', 'M'),
    !Sample %in% c('Doped Oligo', 'Standard Oligo', 'Standard Oligo (ErrASE)')
  ) %>%
  mutate(Diff = factor(Diff)) %>%
  group_by(Sample, Treatment, Type) %>%
  do(tidy(kruskal.test(Fold_2 ~ Diff, data=.))) %>%
  ungroup() %>%
  select(Sample, Treatment, Type, statistic, p.value)

# transition vs transversions
enzPref.idm %>%
  filter(
    Type == 'M',
    !Sample %in% c('Doped Oligo', 'Standard Oligo', 'Standard Oligo (ErrASE)')
  ) %>%
  group_by(Sample, Treatment) %>%
  do(tidy(wilcox.test(Fold_2 ~ Class, data=.))) %>%
  ungroup() %>%
  select(Sample, Treatment, statistic, p.value)

# median values for paper
enzPref.idm %>%
  filter(Treatment == '2') %>%
  group_by(Sample, Type, Diff) %>%
  summarise(med = median(Fold)) %>%
  spread(Sample, med)
```

```{r, errase_pref}
# ErrASE cg/gc, ag/tc
errase.pref <- enzPref.idm %>%
  filter(
    Sample == 'ErrASE',
    Type == 'M'
  ) %>%
  mutate(
    GC = if_else(Diff == 'GC' | Diff == 'CG', 'GC', 'No'),
    AG = if_else(Diff == 'AG' | Diff == 'TC', 'AG', 'No')
  ) %>%
  select(Pos, Diff, Treatment, Fold, GC, AG) %>%
  gather(Test, Val, GC, AG)

# p.val invariant of fold vs fold_2
errase.pref %>%
  group_by(Treatment, Test) %>%
  do(tidy(wilcox.test(Fold ~ Val, data=.))) %>%
  ungroup() %>%
  select(Test, Treatment, statistic, p.value)

errase.pref %>%
  group_by(Test, Treatment, Val) %>%
  summarise(med = median(Fold))
```


```{r, t7_pref}
# T7 ta/at, cg/gc(ligase), ag/tc
t7.pref <- enzPref.idm %>%
  filter(
    str_detect(Sample, 'T7'),
    Type == 'M'
  ) %>%
  mutate(
    AT = if_else(Diff == 'AT' | Diff == 'TA', 'AT', 'No'),
    GC = if_else(Diff == 'GC' | Diff == 'CG', 'GC', 'No'),
    AG = if_else(Diff == 'AG' | Diff == 'TC', 'AG', 'No')
  ) %>%
  select(Sample, Pos, Diff, Treatment, Fold, AT, GC, AG) %>%
  gather(Test, Val, AT, GC, AG)

t7.pref %>%
  group_by(Sample, Treatment, Test) %>%
  do(tidy(wilcox.test(Fold ~ Val, data=.))) %>%
  ungroup() %>%
  select(Sample, Test, Treatment, statistic, p.value)

t7.pref %>%
  group_by(Sample, Test, Treatment, Val) %>%
  summarise(med = median(Fold))
```

```{r, muts_pref}
# muts ag/tc cg/gc
muts.pref <- enzPref.idm %>%
  filter(
    str_detect(Sample, 'MutS'),
    Type == 'M'
  ) %>%
  mutate(
    GC = if_else(Diff == 'GC' | Diff == 'CG', 'GC', 'No'),
    AG = if_else(Diff == 'AG' | Diff == 'TC', 'AG', 'No')
  ) %>%
  select(Sample, Pos, Diff, Treatment, Fold, GC, AG) %>%
  gather(Test, Val, GC, AG)

muts.pref %>%
  group_by(Sample, Treatment, Test) %>%
  do(tidy(wilcox.test(Fold ~ Val, data=.))) %>%
  ungroup() %>%
  select(Sample, Test, Treatment, statistic, p.value)

muts.pref %>%
  group_by(Sample, Test, Treatment, Val) %>%
  summarise(med = median(Fold))
```

#### Sup - Deletion Preferences
Enzyme specificties for single base deletions
```{r, Enzyme_Prefs_Deletions}
enzPref.idm %>%
  filter(
    Type == 'D',
    !Sample %in% c('Doped Oligo', 'Standard Oligo (ErrASE)')
  ) %>%  ggplot(aes(x=Diff, y=Fold_2, color=Treatment)) +
  geom_boxplot(
    outlier.shape = NA,
    show.legend = FALSE
  ) +
  geom_point(
    position=position_jitterdodge(),
    size = 0.5,
    alpha = 0.8
  ) +
  facet_wrap(~ Sample, ncol=3, scales='free_y') +
  theme(
    axis.title.x=element_blank(),
    legend.title=element_text(size=rel(2)),
    legend.position=c(0.85, 0.10)
    ) +
  guides(colour = guide_legend(override.aes = list(size=5))) +
  scale_color_manual(
    name = 'Treatment Round',
    values = c('#ca0020', '#0571b0')
  ) +
  labs(
    title='Deletion Reduction Relative to Doped Assembly',
       y='Log2-fold Error Rate Change'
  )
```

#### Sup - Insertion Preferences
Enzyme specificties for single base insertions
```{r, Enzyme_Prefs_Insertions}
enzPref.idm %>%
  filter(
    Type == 'I',
    !Sample %in% c('Doped Oligo', 'Standard Oligo (ErrASE)')
  ) %>%  ggplot(aes(x=Diff, y=Fold_2, color=Treatment)) +
  geom_boxplot(
    outlier.shape = NA,
    show.legend = FALSE
  ) +
  geom_point(
    position=position_jitterdodge(),
    size = 0.5,
    alpha = 0.8
  ) +
  facet_wrap(~ Sample, ncol=3, scales='free_y') +
  theme(
    axis.title.x=element_blank(),
    legend.title=element_text(size=rel(2)),
    legend.position=c(0.85, 0.10)
    ) +
  guides(colour = guide_legend(override.aes = list(size=5))) +
  scale_color_manual(
    name = 'Treatment Round',
    values = c('#ca0020', '#0571b0')
  ) +
  labs(
    title='Insertion Reduction Relative to Doped Assembly',
       y='Log2-fold Error Rate Change'
    )
```


# Extra Supplement
## Doped Oligo
### Non-Doped w/ ErrASE
Can we figure out what the noise floor is for our method? Is there a difference between the standard oligo and its error corrected counterpart?
```{r, nonDoped_vs_ErrASE}
allSamps %>%
  filter(Sample %in% c('1_nonDoped', '1_ErrASE-nonDoped', '2_ErrASE-nonDoped')) %>%
  DistribUncert2() %>%
  count(Sample, Type, Pos, wt=FracCount) %>%
  ungroup() %>%
  left_join(readCounts, by='Sample') %>%
  mutate(
    Norm = n / Reads,
    Treatment = str_sub(Sample, 1, 1),
    Sample = str_sub(Sample, 3),
    Sample = Sample %>%
      recode(`ErrASE-nonDoped`='ErrASE',
             nonDoped='Standard Oligo'),
    Type = Type %>%
      factor(levels = c('M', 'I', 'D', 'P', 'S')) %>%
      recode(D = 'Single Base Deletions', I = 'Single Base Insertions',
             P = 'Multiple Base Deletions', S = 'Multiple Base Insertions',
             M = 'Mismatches')
  ) %T>%
  {
    group_by(., Type) %>%
      do(with(.,
              tidy(pairwise.wilcox.test(Norm, interaction(Sample, Treatment))))) %>%
      print
  } %>%
  ggplot(aes(x=Sample, y=Norm, color=Treatment, group=Treatment)) +
  facet_wrap(~ Type, ncol=2, scales='free_y') +
  geom_point(
    size=0.75,
    alpha=0.8,
    position=position_jitterdodge()
  ) +
  stat_summary(
    fun.y = median, fun.ymin = median, fun.ymax = median,
    geom = 'crossbar',
    width = 0.5,
    color='black',
    position=position_dodge(width=0.7)
  ) +
  scale_y_log10() +
  annotation_logticks(sides='l') +
  theme(
    legend.title=element_text(size=rel(2)),
    legend.position=c(0.75, 0.15)
  ) +
  guides(colour = guide_legend(override.aes = list(size=5))) +
  scale_color_manual(
    name = 'Treatment',
    values=c("#ca0020", "#0571b0")
  ) +
  labs(x = 'Position', y = 'Error Rate')

```

```{r, nonDoped_vs_ErrASE_2}
allSamps %>%
  filter(Sample %in% c('1_nonDoped', '1_ErrASE-nonDoped', '2_ErrASE-nonDoped')) %>%
  filter(Type == 'M') %>%
  count(Sample, Pos, Diff) %>%
  ungroup() %>%
  left_join(readCounts, by = 'Sample') %>%
  mutate(
    Norm = n / Reads,
    Class= Diff %>%
      recode(AT='Transversion', AG='Transition', AC='Transversion',
             TA='Transversion', TG='Transversion', TC='Transition',
             GA='Transition', GT='Transversion', GC='Transversion',
             CA='Transversion', CT='Transition', CG='Transversion')
  ) %>%
  ggplot(aes(x = Diff, y = Norm, color=Class)) +
  geom_boxplot(outlier.shape = NA, show.legend = FALSE) +
  geom_jitter(position=position_jitter(w = 0.5), size=0.75, alpha=0.8) +
  facet_wrap(~ Sample, ncol = 1) +
  labs(
    y = 'Error Rate',
    title = 'Mismatch Sub-types per Pos.'
  ) +
  theme(
    legend.position='bottom',
    legend.key.size=unit(0.75, "cm"),
    axis.title.x=element_blank(),
    plot.title = element_text(size=rel(1.75))
  ) +
  scale_y_continuous(labels = scientific_format()) +
  scale_color_manual(values=c("#ca0020", "#0571b0")) +
  guides(colour = guide_legend(override.aes = list(size=5)))


```

### Doped Analysis
Here we will run the same sorts of analysis as Figure 2 on the Doped oligo.
```{r, doped_analysis, fig.width=17.5, fig.height=17.5}
positions_dope <- doped %>%
  DistribUncert2() %>%
  count(Pos, Type, wt=FracCount) %>%
  filter(Type != 'S') %>%
  ungroup() %>%
  mutate(
    Norm = n / subset(readCounts, Sample == '1_nonDoped')$Reads,
    Type = Type %>%
      factor(levels = c('M', 'I', 'D', 'P')) %>%
      recode(D = 'Single Base Deletions', I = 'Single Base Insertions',
             P = 'Multiple Base Deletions', M = 'Mismatches')
  ) %>%
  ggplot(aes(x=Pos, y=Norm)) +
  geom_point(size=3) +
  facet_wrap(~ Type, ncol=2) +
  stat_smooth(se=F) +
  labs(x = 'Position',
       y = 'Error Rate',
       title = 'Error Rate vs. Position for Error Sub-types') +
  scale_y_log10(labels = scientific_format()) +
  annotation_logticks(sides='l')

#-------------------------------------------------------------------------------
# Panel 1b
# Percentage of Error Subtypes

sub_type_dope <- doped %>%
  DistribUncert2() %>%
  count(Type, wt=FracCount) %>%
  mutate(
    Norm = n / sum(n) * 100,
    Type = Type %>%
      factor(levels = c('M', 'D', 'P', 'I', 'S'))  %>%
      recode(M = 'MM', D = 'Del.', P = 'M. Del.', I = 'Ins.', S = 'M. Ins.')
  ) %T>%
  {arrange(., -Norm) %>% print()} %>%
  ggplot(aes(x=Type, y=Norm)) +
  geom_bar(stat='identity') +
  theme(plot.title = element_text(size=rel(2))) +
  labs(
    y = 'Percent',
    x = 'Error Type',
    title = 'Percentage of Error Sub-types'
  )

#-------------------------------------------------------------------------------
# Panel c
# plot the distribution of total mismatches per position
mm_freq_dope <- doped %>%
  filter(Type == 'M') %>%
  mutate(
    To = str_sub(Diff, 2, 2),
    From = str_sub(Diff, 1, 1)
  ) %>%
  count(Pos, From) %>%
  ungroup() %>%
  mutate(Norm = n / subset(readCounts, Sample == '1_nonDoped')$Reads) %T>%
  { # pairwise wilcox test and print median values for paper
    group_by(., From) %>%
      summarise(med=median(Norm)) %>%
      arrange(-med) %>%
      print; # <- ; critical for . to be interpreted correctly
    with(., pairwise.wilcox.test(n, From)) %>% print
  } %>%
  ggplot(aes(x = From, y = Norm)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(position=position_jitter(w = 0.5), size=0.75, alpha=0.8) +
  # stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom = 'crossbar', width = 0.5) +
  labs(y = 'Error Rate per Base',
       title = 'Mismatches per Pos.') +
  theme(
    axis.text.x = element_text(size=28),
    axis.title.x = element_blank(),
    plot.title = element_text(size=rel(2.25))
  ) +
  scale_y_continuous(labels = scientific_format())

#-------------------------------------------------------------------------------
# Panel d
# what bases are most likely mutated to
# We will normallize by the total count in each "from" group
mm_type_dope <- doped %>%
  filter(Type == 'M') %>%
  count(Pos, Diff) %>%
  ungroup() %>%
  mutate(
    Char=str_sub(Diff, 1, 1),
    Class= Diff %>%
      recode(AT='Transversion', AG='Transition', AC='Transversion',
             TA='Transversion', TG='Transversion', TC='Transition',
             GA='Transition', GT='Transversion', GC='Transversion',
             CA='Transversion', CT='Transition', CG='Transversion')

  ) %>%
  mutate(Norm = n / subset(readCounts, Sample == '1_nonDoped')$Reads) %T>%
  {# significance testing and printing for paper
    group_by(., Diff) %>%
      summarise(med=median(Norm)) %>%
      arrange(-med) %>%
      print; # <- ; critical for . to be interpreted correctly
      with(., pairwise.wilcox.test(Norm, Diff)) %>% print
  } %>%
  ggplot(aes(x = Diff, y = Norm, color=Class)) +
  geom_boxplot(outlier.shape = NA, show.legend = FALSE) +
  geom_jitter(position=position_jitter(w = 0.5), size=0.75, alpha=0.8) +
  # stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom = 'crossbar', width = 0.5, color='black') +
  labs(
    y = 'Error Rate per Base',
    title = 'Mismatch Sub-types per Pos.'
    ) +
  theme(
    legend.position='bottom',
    legend.key.size=unit(0.75, "cm"),
    axis.title.x=element_blank(),
    axis.text.x = element_text(angle = 315, vjust=0.5),
    plot.title = element_text(size=rel(1.75))
  ) +
  scale_y_continuous(labels = scientific_format()) +
  scale_color_manual(values = c('#7b3294', '#008837')) +
  guides(colour = guide_legend(override.aes = list(size=5)))

#-------------------------------------------------------------------------------
# plot everything!
grid.arrange(
  LabelMaker(positions_dope, 'A)'),
  arrangeGrob(
    LabelMaker(sub_type_dope, 'B)'),
    LabelMaker(mm_freq_dope, 'C)'),
    LabelMaker(mm_type_dope, 'D)'),
    nrow=1),
  ncol=1,
  heights = c(1, 0.67)
  )

#-------------------------------------------------------------------------------
# Values for text

# differences in medians in annealing regions and outside
doped %>%
  DistribUncert2() %>%
  count(Type, Pos, wt=FracCount) %>%
  mutate(Norm = n / subset(readCounts, Sample == '1_nonDoped')$Reads) %>%
  filter(Type == 'P') %>%
  mutate(Region = if_else(Pos >= 36 & Pos <=64, 'Anneal', 'No')) %T>%
  {wilcox.test(Norm ~ Region, data=.) %>% print} %>%
  group_by(Region) %>%
  summarise(med=median(Norm), IQR=IQR(Norm))
```

Here we will calculate the per-base error rate
```{r, doped_errRate}
doped %>%
  DistribUncert2() %>%
  count(Sample, Name, Type, wt=FracCount) %>%
  left_join(
    filter(charCounts, Sample == '1_DopedTemp'),
    by=c('Sample', 'Name')
  ) %>%
  group_by(Type) %>%
  summarise(m=sum(n/Len)/filter(readCounts, Sample == '1_DopedTemp')$Reads)
```

### Doped vs Non-Doped Error Rates
Here we compare the doped oligo to the non-doped. We see that all error rates are significantly higher in the doped sample.
```{r, Dope_vs_NonDope}
allSamps %>%
  filter(Sample %in% c('1_DopedTemp', '1_nonDoped')) %>%
  DistribUncert2() %>%
  count(Sample, Type, Pos, wt=FracCount) %>%
  ungroup() %>%
  left_join(readCounts, by='Sample') %>%
  select(., -Errs) %>%
  # count all errors regardless of type
  bind_rows(.,
            count(., Sample, Reads, Pos, wt=n) %>% mutate(Type = 'A') %>% rename(n=nn)) %>%
  mutate(
    Norm = n / Reads,
    Sample = if_else(Sample == '1_DopedTemp',
                     'Error-Doped Oligo',
                     'Standard Oligo'),
    Type = Type %>%
      factor(levels = c('A', 'M', 'D', 'P', 'I', 'S')) %>%
      recode(D = 'Single Base Deletions', I = 'Single Base Insertions',
             P = 'Multiple Base Deletions', S = 'Multiple Base Insertions',
             M = 'Mismatches', A = 'All Errors')
  ) %T>%
  { # significance test and summarise for paper
    group_by(., Sample, Type) %>%
      summarise(med=median(Norm), mean=mean(Norm)) %>%
      arrange(Sample) %>%
      print();
    group_by(., Type) %>%
      summarise(p.val=wilcox.test(Norm ~ Sample, data=.)$p.value) %>%
        print()
    } %>%
  ggplot(aes(x=Sample, y=Norm)) +
  geom_jitter(position=position_jitter(w = 0.35)) +
  stat_summary(fun.y = median, fun.ymin = median, fun.ymax = median, geom = 'crossbar', width = 0.75) +
  facet_wrap(~ Type, scales='free_y', ncol=3) +
  labs(y = 'Error Rate') +
  theme(axis.title.x=element_blank())
```

# Misc
## MutS Mismatch Preferences
There's some really strange bi-modal correction of CA mismatches with MutS across both samples. Could be guanine oxidization as suggested by Potapov 2017.

```{r, MutS_CA}
enzPref.idm %>%
  filter(
    str_detect(Sample, 'MutS'),
    Type == 'M',
    Diff == 'CA'
    ) %>%
  ggplot(aes(x=Pos, y=Rel_Norm, color=Treatment)) +
  geom_point(size=5) +
  facet_wrap(~ Sample, ncol=2) +
  guides(colour = guide_legend(override.aes = list(size=5))) +
  theme(
    legend.title = element_text(size=rel(2)),
    legend.position = 'bottom'
  ) +
  scale_color_manual(
    name = 'Treatment',
    values=c("#ca0020", "#0571b0")
  ) +
  labs(x = 'Position', y = 'Log2-fold Error Rate Change')
```

## Aligner Comparision
Here we will compare BBMap, Bowtie2, and our NW aligner
```{r, aligners}
bind_rows(
  list(bbmap=fread('./pipeline/1_nonDoped.bbmap.csv', header=T),
       bowtie=fread('./pipeline/1_nonDoped.bowtie.csv', header=T),
       nw=select(nonDoped, -Sample)),
  .id = 'Aligner'
) %>%
  DistribUncert2() %>%
  count(Aligner, Pos, Type, wt=FracCount) %>%
  ungroup() %>%
  mutate(
    Norm=n / subset(readCounts, Sample == '1_nonDoped')$Reads,
    Type = Type %>%
      factor(levels = c('M', 'I', 'D', 'P', 'S')) %>%
      recode(D = 'Single Base Deletions', I = 'Single Base Insertions',
             P = 'Multiple Base Deletions', S = 'Multiple Base Insertions',
             M = 'Mismatches')) %>%
  ggplot(aes(x=Pos, y=Norm, color=Aligner)) +
  facet_wrap(~ Type, ncol=2, scales='free_y') +
  geom_point() +
  stat_smooth(se=F) +
  theme(
    legend.title=element_text(size=rel(2)),
    legend.position=c(0.75, 0.15)
  ) +
  guides(colour = guide_legend(override.aes = list(size=5))) +
  scale_color_discrete(name = 'Aligner',
                       labels = c('BBMap', 'Bowtie2', 'Needleman-Wunsch')) +
  labs(x = 'Position', y = 'Error Rate') +
  scale_y_continuous(labels = scientific_format())
```

## Classic Table
```{r, classic_table}
single <- allSamps %>%
    filter(!Type %in% c('S', 'P')) %>%
      DistribUncert2() %>%
      group_by(Sample, Type, Diff) %>%
      summarise(n=sum(FracCount)) %>%
      ungroup()

# just count the number of multiple counts
multiple <- allSamps %>%
  filter(Type %in% c('S', 'P')) %>%
  group_by(Sample, Type) %>%
  summarise(n=n(), Diff='N/A') %>%
  ungroup()

# grab transitions/transversions
trans <- single %>%
  filter(Type == 'M') %>%
  mutate(
    Type= Diff %>%
      recode(AT='Transversion', AG='Transition', AC='Transversion',
             TA='Transversion', TG='Transversion', TC='Transition',
             GA='Transition', GT='Transversion', GC='Transversion',
             CA='Transversion', CT='Transition', CG='Transversion')
  ) %>%
  group_by(Sample, Type) %>%
  summarise(Diff='N/A', n=sum(n)) %>%
  ungroup()

bind_rows(single, multiple, trans) %>%
  mutate(
    Type = Type %>%
      recode(D = 'Single Base Deletions', I = 'Single Base Insertions',
             P = 'Multiple Base Deletions', S = 'Multiple Base Insertions',
             M = 'Mismatches')
  ) %>%
  spread(Sample, n) %>%
  # manual selection maddness to aid downstream processing
  select(
    Type, Diff,
    `1_nonDoped`, matches('ErrASE-'), `1_DopedTemp`,
    matches('_ErrASE'), contains("1900"), contains("950"),
    contains("Survey"), contains("Furhmann"), `1_T7EndoI`,
    `2_T7EndoI`, contains("e3T7"), contains("e4T7"),
    contains("T4"), contains("EndoV")
  ) %>%
  write.csv('Table_1.csv', quote=FALSE, row.names=FALSE)
```

# Review
## Figure 6
### Error Rates
```{r, review-error-rates}
review.errRate <- review %>%
  DistribUncert2() %>%
  count(Sample, Name, wt=FracCount) %>%
  left_join(charCounts %>% filter(Sample %in% review.samples), ., by = c('Sample', 'Name')) %>%
  replace_na(list(n=0)) %>%
  group_by(Sample) %>%
  summarise(
    errRate = mean(n * (1000/Len)),
    sem = sd(n*(1000/Len)) / sqrt(n())
  ) %>%
  left_join(readCounts, by='Sample') %>%
  mutate(
    PercentPerf = (Reads - Errs) / Reads * 100,
    Construct = str_extract(Sample, "C([023])"),
    Polymerase = str_extract(Sample, "-(.*)-") %>% str_replace_all("-", ""),
    Repeat = str_extract(Sample, "\\d$")
  ) %>%
  group_by(Polymerase, Construct) %>%
  summarise(
    `Error / kb` = mean(errRate),
    `Percent Perfect` = mean(PercentPerf)
    # SEM = mean(sem),
    # PP_SEM = sd(PercentPerf) / n()
  ) %>%
  ungroup() %>%
  gather(Metric, Value, -Polymerase, -Construct)
```

```{r, review-error-rates-plot}
review.errRate.plot <- review.errRate %>%
  # pretty print for fig
  mutate(
    Construct = if_else(Construct == 'C3', 'C1', 'C2'),
    Polymerase = if_else(Polymerase == 'Q5', 'Q5', 'KAPA2G Robust')
  ) %>%
  ggplot(aes(x=Construct, y=Value, fill = Polymerase)) +
  geom_bar(stat='identity', position = 'dodge') +
  facet_wrap(~ Metric, scales = 'free_y') +
  scale_fill_manual(
    name = 'Polymerase',
    values = c('#d8b365','#5ab4ac')
  ) +
  theme(
    legend.position = 'bottom',
    plot.title = element_text(size = rel(2)),
    axis.title.y = element_blank()
  ) +
  labs(title = 'Assembly Qualtiy vs. Polymerase')
```

### Mismatch Sub-Type Preference
```{r, review-mm-subtype}
review.mm.subtype <- review %>%
  filter(Type == 'M') %>%
  count(Sample, Pos, Diff) %>%
  ungroup() %>%
  mutate(
    Char=str_sub(Diff, 1, 1),
    Construct = str_extract(Sample, "C([23])"),
    Polymerase = str_extract(Sample, "-(.*)-") %>% str_replace_all("-", ""),
    Repeat = str_extract(Sample, "\\d$"),
    Class=Diff %>%
      recode(AT='Transversion', AG='Transition', AC='Transversion',
             TA='Transversion', TG='Transversion', TC='Transition',
             GA='Transition', GT='Transversion', GC='Transversion',
             CA='Transversion', CT='Transition', CG='Transversion')
  ) %>%
  inner_join(review.counts, by = c('Construct', 'Char')) %>%
  inner_join(readCounts, by = 'Sample') %>%
  group_by(Construct, Polymerase, Diff, Pos) %>%
  summarise(
    Norm = mean(n / Count / Reads),
    Class = Class
  ) %>%
  ungroup()
```

```{r, review-mm-subtype-plot}
review.subtype.plot <- review.mm.subtype %>%
  # pretty print for fig
  mutate(
    Construct = if_else(Construct == 'C3', 'C1', 'C2'),
    Polymerase = if_else(Polymerase == 'Q5', 'Q5', 'KAPA2G Robust')
  ) %>%
  ggplot(aes(x=Diff, y=Norm, color=Class)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(position = position_jitter(w=0.3), size = 0.5, alpha = 0.8) +
  facet_grid(Construct ~ Polymerase) +
  scale_y_log10() +
  annotation_logticks(sides = 'l') +
  scale_color_manual(values = c('#7b3294', '#008837')) +
  # guides(colour = guide_legend(override.aes = list(size=5))) +
  theme(
    legend.position = 'bottom',
    axis.text.x = element_text(size = rel(0.75))
  ) +
  labs(
    y = 'Error Rate',
    title = 'Mismatch Sub-types per Pos.'
  )

# test transitions vs transversions
review.mm.subtype %>%
  group_by(Construct, Polymerase) %>%
  do(tidy(wilcox.test(Norm ~ Class, data=.))) %>%
  select(-method)

review.mm.subtype %>%
  group_by(Polymerase, Class, Construct) %>%
  summarise(Median = median(Norm)) %>%
  summarise(Mean = mean(Median))

# test Q5 C/G -> A/T vs transitions
review.mm.subtype %>%
  filter(Class == 'Transition' | Diff %in% c('CA', 'GT')) %>%
  mutate(Class = if_else(Class == 'Transversion', 'C/G -> A/T', Class)) %>%
  group_by(Construct, Polymerase, Class) %>%
  summarise(Med = median(Norm), N = n())

review.mm.subtype %>%
  filter(Class == 'Transition' | Diff %in% c('CA', 'GT')) %>%
  mutate(Class = if_else(Class == 'Transversion', 'C/G -> A/T', Class)) %>%
  group_by(Construct, Polymerase) %>%
  do(tidy(wilcox.test(Norm ~ Class, data=.))) %>%
  select(-method)
```

### Overlap Effects
Are the decreases in multiple base deletions in the overlaps real?

```{r, review-overlaps}
overlaps <- c(seq(41, 60), seq(81, 100), seq(121, 140), seq(161, 180))
primers <- c(seq(1, 15), seq(206, 220))

overlap.data <- review %>%
  filter(Type == 'P') %>%
  count(Sample, Pos) %>%
  inner_join(readCounts, by = 'Sample') %>%
  mutate(
    Construct = str_extract(Sample, "C([23])"),
    Polymerase = str_extract(Sample, "-(.*)-") %>% str_replace_all("-", ""),
    Repeat = str_extract(Sample, "\\d$")
  ) %>%
  group_by(Construct, Polymerase, Pos) %>%
  summarise(Norm = mean(n / Reads)) %>%
  ungroup() %>%
  mutate(
    Overlap = case_when(Pos %in% primers ~ 'Primers',
                        Pos %in% overlaps ~ 'Overlaps',
                        TRUE ~ 'Other')
  )
```

```{r, review-overlap-plot}
overlap.plot <- overlap.data %>%
  # better naming for figure
  mutate(
    Construct = if_else(Construct == 'C3', 'C1', 'C2'),
    Polymerase = if_else(Polymerase == 'Q5', 'Q5', 'KAPA2G Robust')
  ) %>%
  ggplot(aes(x = Overlap, y = Norm, color=Polymerase)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(position = position_jitterdodge(), size = 0.5, alpha = 0.8) +
  facet_wrap(~ Construct) +
  scale_y_log10() +
  annotation_logticks(sides = 'l') +
  scale_color_manual(
    name = 'Polymerase',
    values = c('#d8b365','#5ab4ac')
  ) +
  theme(
    legend.position = 'bottom',
    plot.title = element_text(size = rel(1.75)),
    axis.text.x = element_text(size = rel(0.65))
  ) +
  labs(
    y = 'Error Rate',
    x = 'Region',
    title = 'Multiple Base Deletions per Base'
  )

# hypothesis test
overlap.data %>%
  group_by(Construct, Polymerase) %>%
  do(tidy(pairwise.wilcox.test(Norm, Overlap)))

overlap.data %>%
  filter(Overlap == 'Primers') %>%
  group_by(Construct) %>%
  do(tidy(wilcox.test(Norm ~ Polymerase, data=.)))

# median values
overlap.data %>%
  group_by(Construct, Polymerase, Overlap) %>%
  summarise(Median = median(Norm))

# diff between medians
overlap.data %>%
  group_by(Construct, Polymerase, Overlap) %>%
  summarise(Median = median(Norm)) %>%
  ungroup() %>%
  spread(Overlap, Median) %>%
  gather(Region, Value, Overlaps, Primers) %>%
  mutate(Rate = Other / Value) %>%
  group_by(Polymerase, Region) %>%
  summarise(Mean = mean(Rate), sd = sd(Rate))
```

### Error Rate Table
Here we calculate the SEM in a more explicit fashion than above. Note that the left_join followed by a NA -> 0 above only really serves as a hack to make the built in mean/sd functions calculate the appropriate number of samples.
```{r, review-table}
review.table <- nonDoped %>%
  bind_rows(review) %>%
  DistribUncert2() %>%
  count(Sample, Type, Name, wt=FracCount) %>%
  inner_join(charCounts, by = c('Sample', 'Name')) %>%
  summarise(
    Sum = sum(n * (1000/Len)),
    SumSq = sum((n * (1000/Len))^2)
  ) %>%
  inner_join(readCounts, by = 'Sample') %>%
  mutate(
    errRate = Sum / Reads,
    sd = sqrt((SumSq - (Sum^2 / Reads)) / (Reads - 1)),
    sem = sd / sqrt(Reads),
    Sample = recode(Sample, `1_nonDoped` = 'C0-Taq*-1'),
    Construct = str_extract(Sample, "C([023])"),
    Polymerase = str_extract(Sample, "-(.*)-") %>% str_replace_all("-", ""),
    Repeat = str_extract(Sample, "\\d$")
  ) %>%
  group_by(Polymerase, Type) %>%
  summarise(
    `Error Rate` = mean(errRate),
    SEM = mean(sem)
  ) %>%
  ungroup()

review.table %>%
  mutate(
    Value = paste(round(`Error Rate`, 4), "±", round(SEM, 4)),
    Type = Type %>%
      factor(levels = c('M', 'D', 'I', 'P', 'S')) %>%
      recode(D = 'Single Base Deletions', I = 'Single Base Insertions',
             P = 'Multiple Base Deletions', M = 'Mismatches',
             S = 'Multiple Base Insertions')
    ) %>%
  select(Polymerase, Type, Value) %>%
  spread(Polymerase, Value)
```

### Actual Figure
```{r, Figure_5}
grid.arrange(
  arrangeGrob(
    LabelMaker(review.errRate.plot, 'A)'),
    LabelMaker(overlap.plot, 'C)'),
    ncol = 1
  ),
  LabelMaker(review.subtype.plot, 'B)'),
  nrow = 1,
  widths = c(0.55, 1)
)
```

## Suplement
### Correlation between Repeats
```{r, corrs}
review %>%
  DistribUncert2() %>%
  count(Sample, Pos, wt = FracCount) %>%
  ungroup() %>%
  inner_join(readCounts, by = 'Sample') %>%
  mutate(
    Norm = n / Reads,
    Construct = str_extract(Sample, "C([23])"),
    Polymerase = str_extract(Sample, "-(.*)-") %>% str_replace_all("-", ""),
    Repeat = str_extract(Sample, "\\d$")
  ) %>%
  select(-Sample, -Reads, -Errs, -n) %>%
  spread(Repeat, Norm) %T>%
  {
    group_by(., Construct, Polymerase) %>%
      summarise(Cor = cor(`1`, `2`)) %>%
      print()
  } %>%
  # pretty plot
  mutate(
    Construct = if_else(Construct == 'C3', 'C1', 'C2'),
    Polymerase = if_else(Polymerase == 'Q5', 'Q5', 'KAPA2G Robust')
  ) %>%
  ggplot(aes(x=`1`, y=`2`)) +
  geom_point() +
  geom_abline() +
  facet_grid(Polymerase ~ Construct) +
  scale_x_log10() +
  scale_y_log10() +
  annotation_logticks() +
  labs(
    x = 'Techical Repeat 1',
    y = 'Technical Repeat 2'
  )
```

### Classic Positional Graph

```{r, review-errs-pos}
review %>%
  DistribUncert2() %>%
  count(Sample, Type, Pos, wt = FracCount) %>%
  ungroup() %>%
  mutate(
    Construct = str_extract(Sample, "C([23])"),
    Polymerase = str_extract(Sample, "-(.*)-") %>% str_replace_all("-", ""),
    Repeat = str_extract(Sample, "\\d$")
  ) %>%
  inner_join(readCounts, by = 'Sample') %>%
  group_by(Construct, Polymerase, Type, Pos) %>%
  summarise(Norm = mean(n / Reads)) %>%
  # better naming for figure
  ungroup() %>%
  mutate(
    Construct = if_else(Construct == 'C3', 'C1', 'C2'),
    Polymerase = if_else(Polymerase == 'Q5', 'Q5', 'KAPA2G Robust'),
    Type = Type %>%
      factor(levels = c('M', 'D', 'P', 'I', 'S'))  %>%
      recode(M = 'MM', D = 'Del.', P = 'M. Del.', I = 'Ins.', S = 'M. Ins.')
  ) %>%
  ggplot(aes(x=Pos, y=Norm, color=Polymerase)) +
  geom_point() +
  geom_smooth(se = F, method = 'loess') +
  facet_grid(Type ~ Construct) +
  scale_y_log10() +
  annotation_logticks(sides = 'l') +
  scale_color_manual(
    name = 'Polymerase',
    values = c('#d8b365','#5ab4ac')
  ) +
  theme(legend.position = 'bottom') +
  labs(
    x = 'Position',
    y = 'Error Rate'
  )
```

### Mismatch Preference
```{r, review-mm}
review.mm <- review %>%
  filter(Type == 'M') %>%
  mutate(
    To = str_sub(Diff, 2, 2),
    From = str_sub(Diff, 1, 1)
  ) %>%
  count(Sample, Pos, From) %>%
  ungroup() %>%
  mutate(
    Construct = str_extract(Sample, "C([23])"),
    Polymerase = str_extract(Sample, "-(.*)-") %>% str_replace_all("-", ""),
    Repeat = str_extract(Sample, "\\d$")
  ) %>%
  inner_join(rename(review.counts, From=Char), by = c('Construct', 'From')) %>%
  inner_join(readCounts, by = 'Sample') %>%
  group_by(Construct, Polymerase, From, Pos) %>%
  summarise(Norm = mean(n / Count / Reads)) %>%
  ungroup()

# plot it
review.mm %>%
  ggplot(aes(x=From, y=Norm, color=Construct)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(position = position_jitter(w=0.3), size = 0.5, alpha = 0.8) +
  facet_grid(Polymerase ~ Construct, scales='free_y') +
  scale_y_log10() +
  annotation_logticks(sides = 'l')

# significance testing for A/T or G/C broadly
review.mm %>%
  mutate(Class = if_else(From %in% c('A', 'T'), 'A/T', 'C/G')) %>%
  group_by(Construct, Polymerase, Class) %>%
  summarise(Median = median(Norm))

review.mm %>%
  mutate(Class = if_else(From %in% c('A', 'T'), 'A/T', 'C/G')) %>%
  group_by(Construct, Polymerase) %>%
  do(tidy(wilcox.test(Norm ~ Class, data=.))) %>%
  select(-method)

# significance testing for all pairwise comparisons
review.mm %>%
  group_by(Construct, Polymerase, From) %>%
  summarise(Median = median(Norm))

review.mm %>%
  group_by(Construct, Polymerase) %>%
  do(with(., tidy(pairwise.wilcox.test(Norm, From)))) %>%
  print(n=24)
```

### Mismatch Hotspots
```{r, hot-spots}
review %>%
  filter(Type == 'M') %>%
  count(Sample, Pos, Diff) %>%
  ungroup() %>%
  inner_join(readCounts, by = 'Sample') %>%
  mutate(
    Norm = n / Reads, Construct = str_extract(Sample, "C([023])"),
    Polymerase = str_extract(Sample, "-(.*)-") %>% str_replace_all("-", ""),
    Repeat = str_extract(Sample, "\\d$")
    ) %>%
  filter(Polymerase == 'Taq') %>%
  tbl_df %>%
  group_by(Sample) %>%
  top_n(n=6, wt = Norm) %>%
  ungroup() %>%
  filter(Construct == 'C2') %>%
  arrange(-Norm)
```

### Polymerase Slippage
```{r, pol-slip}
review %>%
  filter(Type == 'S') %>%
  count(Sample, Pos, Diff) %>%
  ungroup() %>%
  inner_join(readCounts, by = 'Sample') %>%
  mutate(
    Norm = n / Reads, Construct = str_extract(Sample, "C([023])"),
    Polymerase = str_extract(Sample, "-(.*)-") %>% str_replace_all("-", ""),
    Repeat = str_extract(Sample, "\\d$")
    ) %>%
  filter(Polymerase == 'Taq', Construct == 'C2') %>%
  tbl_df %>%
  group_by(Sample) %>%
  top_n(n=6, wt = Norm) %>%
  ungroup() %>%
  arrange(-Norm)
```

### Are Mismatches the Last Base Snythesized
```{r, last-base}
last.base <- c(c2, c3) %>%
  str_split("") %>%
  unlist() %>%
  data.table(
    LastBase = .,
    Construct = c(rep("C2", 220), rep("C3", 220)),
    Pos = rep(seq(2, 221), 2)
  )

last.base.test <- review.mm.subtype %>%
  select(-Class) %>%
  inner_join(last.base, by = c('Construct', 'Pos')) %>%
  mutate(
    From = str_sub(Diff, 1, 1),
    To = str_sub(Diff, 2, 2)
  ) %>%
  filter(!LastBase == From) %>%
  mutate(Remain = if_else(LastBase == To, 'Yes', 'No'))

# plot it
last.base.test %>%
  ggplot(aes(x=Remain, y=Norm)) +
  geom_boxplot() +
  facet_grid(Polymerase ~ Construct, scales='free_y') +
  scale_y_log10() +
  annotation_logticks(sides='l')

# doesnt look significant, but to be sure...
last.base.test %>%
  group_by(Construct, Polymerase) %>%
  do(tidy(wilcox.test(Norm ~ Remain, data=.)))
```
